parquet-hadoop/src/main/java/org/apache/parquet/hadoop/DirectCodecPool.java

```diff
@@ -0,0 +1,239 @@
+    }
+    DIRECT_DECOMPRESSION_CODEC_CLASS = tempClass;
+    CREATE_DIRECT_DECOMPRESSOR_METHOD = tempCreateMethod;
+    DECOMPRESS_METHOD = tempDecompressMethod;
+  }
+
+  private DirectCodecPool() {
+  }
+
+  public class CodecPool {
+    private final GenericObjectPool compressorPool;
+    private final GenericObjectPool decompressorPool;
+    private final GenericObjectPool directDecompressorPool;
+    private final boolean supportDirectDecompressor;
+
+    private CodecPool(final CompressionCodec codec){
+      try {
+        boolean supportDirectDecompressor = codec.getClass() == DIRECT_DECOMPRESSION_CODEC_CLASS;
+        compressorPool = new GenericObjectPool(new BasePoolableObjectFactory() {
+          public Object makeObject() throws Exception {
+            return codec.createCompressor();
+          }
+        }, Integer.MAX_VALUE);
+
+        Object com = compressorPool.borrowObject();
+        if (com != null) {
+          cPools.put(com.getClass(), compressorPool);
+          compressorPool.returnObject(com);
+        } else {
+          if (Log.WARN) {
+            LOG.warn("Unable to find compressor for codec " + codec.getClass().getName());
+          }
+        }
+
+        decompressorPool = new GenericObjectPool(new BasePoolableObjectFactory() {
+          public Object makeObject() throws Exception {
+            return codec.createDecompressor();
+          }
+        }, Integer.MAX_VALUE);
+
+        Object decom = decompressorPool.borrowObject();
+        if (decom != null) {
+          dePools.put(decom.getClass(), decompressorPool);
+          decompressorPool.returnObject(decom);
+        } else {
+          if (Log.WARN) {
+            LOG.warn("Unable to find decompressor for codec " + codec.getClass().getName());
+          }
+        }
+
+        if (supportDirectDecompressor) {
+          directDecompressorPool = new GenericObjectPool(
+              new BasePoolableObjectFactory() {
+            public Object makeObject() throws Exception {
+              return CREATE_DIRECT_DECOMPRESSOR_METHOD.invoke(DIRECT_DECOMPRESSION_CODEC_CLASS);
+            }
+          }, Integer.MAX_VALUE);
+
+          Object ddecom = directDecompressorPool.borrowObject();
+          if (ddecom != null) {
+            directDePools.put(ddecom.getClass(), directDecompressorPool);
+            directDecompressorPool.returnObject(ddecom);
+
+          } else {
+            supportDirectDecompressor = false;
+            if (Log.WARN) {
+              LOG.warn("Unable to find direct decompressor for codec {}" + codec.getClass().getName());
+            }
+          }
+
+        } else {
+          directDecompressorPool = null;
+        }
+
+        this.supportDirectDecompressor = supportDirectDecompressor;
+      } catch (Exception e) {
+        throw new ParquetRuntimeException("Error creating compression codec pool.") {
+          // This exception will not likely be recoverable
```

Comment:
done, also removed anonymous exception class and replaced it with ParquetCompressionCodecException